### DISCUSSION TITLE
How to inject singleton style tag into a closed shadow root

### DISCUSSION THREAD

**Discussion by veesahni** [webpack/webpack#14714] [Q&A]
Given:
* a single closed shadow-root that encapsulates multiple components
* a tree of components with colocated css files, where each imports styles to use locally i.e. ```import styles from './something.css'```

I want to:
* inject a singleton style tag into the shadow root

Thoughts:
* style-loader has an insert option which could be given a function that inserts into the shadow root.  But since the shadow root is closed, only the component that created the shadow root has a reference to it.
* There doesn't seem to be a documented way to "import" the style-loader's final generated CSS at runtime by the component that is managing the shadow root

Here's an ugly workaround:

Create a runtime object that aggregates the CSS provided by css-loader and then have the aggregator provide the concatenated string to the top level component when it needs it in its render:

```js
import aggregator from './aggregator.js'
import sheet from './something.css'

const styles = sheet.locals
aggregator.append(sheet.toString())

/* rest of component */
```

it works, but it requires adding extra lines into every single component to aggregate the CSS.

**Is there a cleaner way to do this?**

**Comment by alexander-akait** [comment]
You can't inject singleton in shadow root

**Reply by alexander-akait** [reply]
Here right solution https://github.com/webpack-contrib/style-loader#custom-elements-shadow-dom

**Reply by veesahni** [reply]
@alexander-akait Appreciate the response.  I did see the lazyStyleTag solution, but it has two challenges:
* Every component that has it's own styles now has to retrieve a reference to the shadowRoot to send to .use() .. so this is similar extra code as the workaround I suggested
* style-loader will update the DOM n times instead of 1 time, creating more work for the browser .. which is sub-optimal for production use.

.. but I guess its the best we have from officially documented options at the moment.
